library/Icingadb/Common/IcingaRedis.php

```diff
@@ -69,60 +65,118 @@ public function getIcingaRedis()
+            } else {
+                $fields[$key] = $kv;
+                $key = null;
+            }
+        }
 
-        return $stats['timestamp'] / 1000;
+        return $fields['timestamp'] / 1000;
     }
 
-    private function getPrimaryRedis()
+    private static function getPrimaryRedis(Config $config = null)
     {
-        $config = Config::module('icingadb')->getSection('redis1');
-        $redis = new Redis();
+        if ($config === null) {
+            $config = Config::module('icingadb');
+        }
+
+        $section = $config->getSection('redis1');
 
-        $redis->connect(
-            $config->get('host', 'localhost'),
-            $config->get('port', 6380),
-            0.5
-        );
+        $redis = new Redis([
+            'host'      => $section->get('host', 'localhost'),
+            'port'      => $section->get('port', 6380),
+            'timeout'   => 0.5
+        ] + static::getTlsParams($config));
+
+        $redis->ping();
 
         return $redis;
     }
 
-    private function getSecondaryRedis()
+    private static function getSecondaryRedis(Config $config = null)
     {
-        $config = Config::module('icingadb')->getSection('redis2');
-        $redis = new Redis();
+        if ($config === null) {
+            $config = Config::module('icingadb');
+        }
 
-        $host = $config->host;
+        $section = $config->getSection('redis2');
+        $host = $section->host;
 
         if (empty($host)) {
             return null;
         }
 
-        $redis->connect(
-            $host,
-            $config->get('port', 6380),
-            0.5
-        );
+        $redis = new Redis([
+            'host'      => $host,
+            'port'      => $section->get('port', 6380),
+            'timeout'   => 0.5
+        ] + static::getTlsParams($config));
+
+        $redis->ping();
 
         return $redis;
     }
 
-    private static function streamSupportAvailable()
+    private static function getTlsParams(Config $config)
+    {
+        $config = $config->getSection('redis');
+
+        if (! $config->get('tls', false)) {
+            return [];
+        }
+
+        $ssl = [];
+        $ca = $config->get('insecure') ? null : $config->get('ca');
+
+        if ($ca === null) {
+            $ssl['verify_peer'] = false;
+            $ssl['verify_peer_name'] = false;
+        } else {
+            $ssl['cafile'] = static::ensurePemOnDisk($ca);
+        }
+
+        $cert = $config->get('cert');
+        $key = $config->get('key');
+
+        if ($cert !== null && $key !== null) {
+            $ssl['local_cert'] = static::ensurePemOnDisk($cert . PHP_EOL . $key);
+        }
+
+        return ['scheme' => 'tls', 'ssl' => $ssl];
+    }
+
+    private static function ensurePemOnDisk($pem)
     {
-        $extRedisVersion = phpversion('redis');
-        return version_compare($extRedisVersion, '4.3.0', '>=');
+        $dir = sys_get_temp_dir() . DIRECTORY_SEPARATOR . sha1($pem);
```

Comment:
Are you sure the certs shall be written and deleted every single time?

Reply:
No, if you are talking about every single redis request. Yes, if we're talking about every user request.

Reply:
Well, it's not just a performance question. The reason why I don't see an issue here, is that we're putting private keys (secrets) onto disk there. But that's also the case if they're part of the ini configuration... This can only be accessed by the owner and group, and so it's the case for the storage dir... (2770 in package installations)

Well, yes, a regular file storage sounds okay? What's your proposal regarding the actual path and structure?

Reply:
Ah, security is the problem! What if I just umask()ify my current implementation?

Reply:
No. We are on a nice track with the storage dir. A perfect chance to use that with a proper use-case and implementation.

Reply:
I'd suggest the framework to allow modules `$STORAGE/modules/$MODULE` and this module to use `$STORAGE/modules/$MODULE/redis/{ca,cert,key}.pem` for saved config and temp files for not yet saved one.

Reply:
Sounds fine to me.